criu/sk-udp.c

```diff
@@ -0,0 +1,655 @@
+		if (ret < 0)
+			goto free_upe;
+		else if (ret == 0) {
+			/* We receive all the packets */
+			xfree(upe);
+			break;
+		}
+
+		aux = ret;
+		upe->pkt_len = ret;
+
+		ret = -1;
+		upe->n_addr = family == AF_INET ? PB_ALEN_INET : PB_ALEN_INET6;
+		upe->addr = xmalloc(pb_repeated_size(upe, addr));
+		if (!upe->addr)
+			goto free_upe;
+		dump_udp_sockaddr(upe, &sa_addr, family);
+
+		up = xmalloc(sizeof(struct udp_packet));
+		if (!up)
+			goto free_addr;
+
+		up->upe = upe;
+		up->data = data;
+		list_add_tail(&up->packet_list, &recv_queue_packets);
+		num_of_pkts += 1;
+		data += aux;
+		pkt_len -= aux;
+	}
+
+	ret = 0;
+	goto out;
+
+free_addr:
+	xfree(upe->addr);
+free_upe:
+	xfree(upe);
+free_data:
+	list_for_each_entry_safe(up, n, &recv_queue_packets, packet_list)
+		free_udp_packet(up);
+	INIT_LIST_HEAD(&recv_queue_packets);
+	xfree(g_data);
+	g_data = NULL;
+out:
+	*inq_pkt = num_of_pkts;
+	return ret;
+}
+
+static int dump_send_queue(struct inet_sk_desc *sk, int family, unsigned int *outq_pkt)
+{
+	int aux = 1, outq_max, ret, addr_len;
+	char *data;
+	UdpPacketEntry *upe;
+	union libsoccr_addr sa_addr;
+
+	*outq_pkt = 0;
+	upe = xmalloc(sizeof(UdpPacketEntry));
+	if (!upe)
+		return -1;
+	udp_packet_entry__init(upe);
+
+	ret = 0;
+	pr_info("Turning UDP repair on for socket %x\n", sk->sd.ino);
+	if (setsockopt(sk->rfd, SOL_UDP, UDP_REPAIR, &aux, sizeof(aux))) {
+		pr_info("Can't turn on UDP repair, skip dumping UDP send queue\n");
+		goto free_upe;
+	}
+
+	aux = 0;
+	if (setsockopt(sk->rfd, SOL_SOCKET, SO_PEEK_OFF, &aux, sizeof(aux))) {
+		/* CentOS 7 does not support SO_PEEK_OFF on AF_INET sockets, the
+		 * peek off is 0 every recvmsg
+		 */
+		if (errno != EOPNOTSUPP) {
+			pr_perror("Can't set SO_PEEK_OFF");
+			goto free_upe;
+		}
+	}
+
+	ret = -1;
+	if (ioctl(sk->rfd, SIOCOUTQ, &outq_max)) {
+		pr_perror("Can't read SIOCOUTQ");
+		goto free_upe;
+	}
+
+	data = xmalloc(outq_max);
+	if (!data)
+		goto free_upe;
+
+	addr_len = family == AF_INET ? sizeof(sa_addr.v4) : sizeof(sa_addr.v6);
+	ret = recv_one_udp(sk->rfd, &sa_addr, addr_len, data, outq_max, NULL, 0);
+	if (ret <= 0)
+		goto free_data;
+
+	*outq_pkt = 1;
+	upe->pkt_len = ret;
+	upe->n_addr = family == AF_INET ? PB_ALEN_INET : PB_ALEN_INET6;
+
+	ret = -1;
+	upe->addr = xmalloc(pb_repeated_size(upe, addr));
+	if (!upe->addr)
+		goto free_data;
+	dump_udp_sockaddr(upe, &sa_addr, family);
+
+	send_queue_packet = xmalloc(sizeof(struct udp_packet));
+	if (!send_queue_packet)
+		goto free_addr;
+
+	send_queue_packet->upe = upe;
+	send_queue_packet->data = data;
+	return 0;
+
+free_addr:
+	xfree(upe->addr);
+free_data:
+	xfree(data);
+free_upe:
+	xfree(upe);
+	*outq_pkt = 0;
+	return ret;
+}
+
+static int udp_lock_one(int fd, struct inet_sk_desc *sk)
+{
+	int ret;
+
+	sk->rfd = dup(fd);
+	if (sk->rfd < 0) {
+		pr_perror("Can't save socket fd for repair");
+		return -1;
+	}
+
+	if (!(root_ns_mask & CLONE_NEWNET)) {
+		ret = nf_lock_connection(sk);
```

Comment:
pls add a comment that explains why we need to block a UDP connection.

Reply:
Looking at this now I think the network lock here is redundant. As with TCP, the network lock prevents the changes in TCP state when a new packet arrives. But with UDP, the packets are quite independent so the lock is not needed. What is your opinion?

Reply:
We may want to lock them only if we want to be sure that we will not see new packets in receive queues. Do we need this?

Reply:
I use normal recv to dump receive queues, if new packets arrive after that recv, they are not dumped. And these new packets does not affect any previous dumped information so I removed the network lock